src/KubernetesClient/Watch.cs

```diff
@@ -0,0 +1,491 @@
+                }
+                return sb.ToString();
+            }
+
+            protected override void Dispose(bool disposing)
+            {
+                base.Dispose(disposing);
+                stream?.Dispose();
+            }
+
+            bool EnsureData()
+            {
+                if (charRead == charWrite)
+                {
+                    try { return EnsureDataAsync().GetAwaiter().GetResult(); }
+                    catch (OperationCanceledException) { return false; } // non-async callers aren't expecting OperationCanceledException
+                }
+                return true;
+            }
+
+            async Task<bool> EnsureDataAsync()
+            {
+                if (charRead == charWrite)
+                {
+                    CancelToken.ThrowIfCancellationRequested();
+                    charRead = charWrite = 0;
+                    do
+                    {
+                        int bytesRead = await stream.ReadAsync(byteBuffer, 0, byteBuffer.Length, CancelToken).ConfigureAwait(false);
+                        if (bytesRead == 0) return false;
+                        charWrite = decoder.GetChars(byteBuffer, 0, bytesRead, charBuffer, 0);
+                    } while(charWrite == 0);
+                }
+                return true;
+            }
+
+            int Read(char[] buffer, int index, int count, bool block)
+            {
+                int start = index;
+                while(count != 0 && EnsureData())
+                {
+                    int toRead = Math.Min(count, charWrite-charRead);
+                    Array.Copy(charBuffer, charRead, buffer, index, toRead);
+                    charRead += toRead;
+                    index += toRead;
+                    if (!block) break;
+                    count -= toRead;
+                }
+                return index - start;
+            }
+
+#if NETCOREAPP2_1
+            async ValueTask<int> ReadAsync(Memory<char> buffer, bool block)
+            {
+                int totalRead = 0;
+                while(buffer.Length != 0 && (charRead != charWrite || await EnsureDataAsync().ConfigureAwait(false)))
+                {
+                    int toRead = Math.Min(buffer.Length, charWrite-charRead);
+                    charBuffer.AsMemory(0, toRead).CopyTo(buffer);
+                    totalRead += toRead;
+                    charRead += toRead;
+                    if (!block) break;
+                    buffer = buffer.Slice(toRead);
+                }
+                return totalRead;
+            }
+#else
+            async Task<int> ReadAsync(char[] buffer, int index, int length, bool block)
+            {
+                int start = index;
+                while(length != 0 && (charRead != charWrite || await EnsureDataAsync().ConfigureAwait(false)))
+                {
+                    int toRead = Math.Min(length, charWrite-charRead);
+                    Array.Copy(charBuffer, 0, buffer, index, toRead);
```

Comment:
`charRead` must be given for `sourceIndex`:
```csharp
Array.Copy(charBuffer, charRead, buffer, index, toRead);
```

Reply:
Shame. Of course you're right.